File: src/main/java/alg01unionfind/UF2QuickUnion.java
package alg01unionfind;


/**
 * <p>
 * Improvement upon quickunion.
 * Union is implemented by setting parent-child nodes
 * thus creating some sort of a tree for complex unions.
 * </p>
 * Defects:
 * <ul>
 * <li>Trees can get tall</li>
 * <li>Finds are too expensive (could be up to N in a worst case)</li>
 * </ul>
 */
public class UF2QuickUnion implements UnionFindAlgorithm {

    protected int[] parentOf;

    public UF2QuickUnion(int size) {
        parentOf = new int[size];
        for (int i = 0; i < size; i++) {
            parentOf[i] = i;
        }
    }

    /**
     * This is the method for finding union tree root.
     * The roots are the nodes, that were never been set as some other's node child
     * thus having the parent it had during the initialization, i.e. i == parentOf[i]
     * being a parent to itself.
     */
    protected int root(int i) {
        while (i != parentOf[i]) {
            // We cycle thought the node's parents until we find a node that is parent
            // to itself
            i = parentOf[i];
        }
        return i;
    }

    /**
     * To union two nodes we set their respective tree roots in as parent-child
     * (who is who does not matter) creating bigger tree in a process
     * This implementation sets N1 as a parent of N2
     */
    @Override
    public void union(int n1, int n2) {
        int rootN1 = root(n1);
        int rootN2 = root(n2);
        parentOf[rootN2] = rootN1;
    }

    /**
     * Here we must find out if the roots of union trees, n1 and n2 are in, are the same
     * If they are -- n1 and n2 are connected and vice versa
     * This could be as slow as O(N) in a worst case
     */
    @Override
    public boolean connected(int n1, int n2) {
        return root(n1) == root(n2);
    }
}

File: src/main/kotlin/alg01unionfind/UF2QuickUnion.kt
package alg01unionfind

/**
 * Fixes the drawbacks of UF1QuickUnion.
 * Union is implemented by setting parent-child nodes creating a tree for complex unions.
 * Drawbacks: trees can get tall making finds expensive (could be up to N in a worst case).
 */
open class UF2QuickUnion(size: Int) : UnionFind {

    protected val parentOf: Array<Int> = Array(size) { index: Int -> index }

    /**
     * This is the method for finding union tree root.
     * The roots are the nodes, that were never been set as an other's node child
     * therefore having the parent it had during the initialization, i.e. i == parentOf[i]
     * meaning it is a parent to itself.
     */
    protected open fun treeRoot(n: Int): Int {
        var root = n
        while (root != this.parentOf[root]) {
            // Cycle thought the node's parents until a node that is parent to itself is found
            root = this.parentOf[root]
        }
        return root
    }

    override fun connected(n1: Int, n2: Int): Boolean {
        return this.treeRoot(n1) == this.treeRoot(n2)
    }

    /**
     * To union two nodes the root of one node is set as a parent to the root of another node
     * (which root is parent to which does not matter) creating a bigger tree in a process.
     * This implementation sets the root of the n1 as a parent to the root of n2.
     */
    override fun union(n1: Int, n2: Int) {
        val n1TreeRoot = this.treeRoot(n1)
        val n2TreeRoot = this.treeRoot(n2)
        parentOf[n2TreeRoot] = n1TreeRoot
    }
}
